### PR TITLE
Rename single keys to the factory parameter 

### PR DESCRIPTION
**Pull request by davidliu**
Avoids naming requirements in the simple case.

Addresses #81.

### CODE REVIEW

File: assisted-inject-sample-android/src/main/java/com/example/Greeter.java [hunk 1]
```diff
@@ -8,8 +8,8 @@
   private final String exclamation;
 
   @AssistedInject
-  Greeter(@Assisted String greeting, @Exclamation String exclamation) {
-    this.greeting = greeting;
+  Greeter(@Assisted String greetg, @Exclamation String exclamation) {
```
Comment (Zhuinden) [inline assisted-inject-sample-android/src/main/java/com/example/Greeter.java]: is this rename intentional? :thinking:

Comment (davidliu) [reply assisted-inject-sample-android/src/main/java/com/example/Greeter.java]: Nope, must have slipped in from testing.

File: assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt [hunk 2]
```diff
@@ -237,21 +237,45 @@ class AssistedInjectProcessor : AbstractProcessor() {
...
         factoryMethod) as ExecutableType
 
-    val expectedKeys = assistedRequests.map { it.namedKey }.sorted()
+    val expectedKeys = assistedRequests.map { it.namedKey }
     val factoryKeys = factoryMethod.parameters
         .zip(factoryExecutable.parameterTypes) { element, mirror -> element.asNamedKey(mirror) }
-    val keys = factoryKeys.sorted()
-    if (keys != expectedKeys) {
+
+    // Rename any single keys to the factory keys to avoid requiring same names
```
Comment (oldergod) [inline assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt]: https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt [hunk 3]
```diff
@@ -237,21 +237,45 @@ class AssistedInjectProcessor : AbstractProcessor() {
...
+    val singleFactoryKeys = factoryKeys
+        .filter { key -> factoryKeys.count { it.key == key.key } == 1 }
+        .toMutableSet()
```
Comment (oldergod) [inline assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt]: Might be more efficient to do group by key and only take unique values.

File: assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt [hunk 4]
```diff
@@ -237,21 +237,45 @@ class AssistedInjectProcessor : AbstractProcessor() {
...
+    val renameableKeys = expectedKeys
+        .map { expectedKey ->
+          val match = singleFactoryKeys.firstOrNull { factoryKey -> factoryKey.key == expectedKey.key }
+          if (match != null) {
+            singleFactoryKeys.remove(match)
```
Comment (oldergod) [inline assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt]: Could you mutate data from outside the `map` ?
`singleFactoryKeys - expectedKeys` should be enough?

Comment (davidliu) [reply assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt]: Mmm, I think this was required during the processing of `expectedKeys` in the following example case:

    class Test {
      @AssistedInject
      Test(@Assisted String fao, @Assisted String faz, String bar) {}

      @AssistedInject.Factory
      interface Factory {
        Test create(String foo);
      }
    }

I'll rewrite this so that it more cleanly only matches single requests against single factory keys.

File: assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt [hunk 5]
```diff
@@ -20,27 +20,14 @@ import com.google.auto.common.MoreTypes
...
-import javax.annotation.processing.ProcessingEnvironment
-import javax.annotation.processing.Processor
-import javax.annotation.processing.RoundEnvironment
+import javax.annotation.processing.*
 import javax.lang.model.SourceVersion
 import javax.lang.model.element.Element
-import javax.lang.model.element.ElementKind.CLASS
-import javax.lang.model.element.ElementKind.CONSTRUCTOR
-import javax.lang.model.element.ElementKind.INTERFACE
+import javax.lang.model.element.ElementKind.*
```
Comment (oldergod) [inline assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt]: no wildcard in imports

**Comment by GuilhE** [comment]
Any idea when it will be merged? Thanks.

**Comment by GuilhE** [comment]
Sorry for being a pain in the ass, but, if all checks have passed, no conflicts, the solution works, what's preventing it to be merged and released?

Let's clean this 😊
```
//    api(Libs.assisted_inject_annotations_dagger2)
compileOnly("com.github.davidliu.AssistedInject:assisted-inject-annotations-dagger2:268bff6") {
        because("Rename single keys to the factory parameter #121")
}
    
//    kapt(Libs.assisted_inject_processor_dagger2)
kapt("com.github.davidliu.AssistedInject:assisted-inject-processor-dagger2:268bff6") {
        because("Rename single keys to the factory parameter #121")
    }

```

Sorry for the presure.

**Comment by janhar** [comment]
We also faced the problem in a multi module project. This PR actualy solves the problem. Thanks to @davidliu 
Just a question, is there any reason why this PR is not accepted yet?

**Comment by oldergod** [comment]
Only ~Square~ oops, Jake knows.

**Comment by GianfrancoMS** [comment]
Any updates?

**Comment by Lampotrias** [comment]
> Sorry for being a pain in the ass, but, if all checks have passed, no conflicts, the solution works, what's preventing it to be merged and released?
> 
> Let's clean this 
> 
> ```
> //    api(Libs.assisted_inject_annotations_dagger2)
> compileOnly("com.github.davidliu.AssistedInject:assisted-inject-annotations-dagger2:268bff6") {
>         because("Rename single keys to the factory parameter #121")
> }
>     
> //    kapt(Libs.assisted_inject_processor_dagger2)
> kapt("com.github.davidliu.AssistedInject:assisted-inject-processor-dagger2:268bff6") {
>         because("Rename single keys to the factory parameter #121")
>     }
> ```
> 
> Sorry for the presure.

I have an error in this case:
`ERROR: Failed to resolve: com.github.davidliu.AssistedInject:assisted-inject-annotations-dagger2:268bff6`

**Comment by davidliu** [comment]
@Lampotrias that's not really related to this PR; check [my comment](https://github.com/square/AssistedInject/issues/81#issuecomment-547094560) in the issue thread for instructions so that we can keep discussion here related to the PR.

@JakeWharton any possibility for updates?

**Comment by GuilhE** [comment]
Any updates?
